executor/executor_runner.h

```diff
@@ -0,0 +1,780 @@
+		state_ = State::Idle;
+#if !SYZ_EXECUTOR_USES_FORK_SERVER
+		if (process_)
+			Restart();
+#endif
+	}
+
+	bool ReadResponse(bool out_of_requests)
+	{
+		uint32 status;
+		ssize_t n = read(resp_pipe_, &status, sizeof(status));
+		if (n == 0) {
+			debug("proc %d: response pipe EOF\n", id_);
+			return false;
+		}
+		if (n != sizeof(status))
+			failmsg("proc resp pipe read failed", "n=%zd", n);
+		if (state_ == State::Handshaking) {
+			debug("proc %d: got handshake reply\n", id_);
+			state_ = State::Idle;
+			Execute();
+		} else if (state_ == State::Executing) {
+			debug("proc %d: got execute reply\n", id_);
+			HandleCompletion(status);
+			if (out_of_requests)
+				wait_start_ = current_time_ms();
+		} else {
+			debug("got data on response pipe in wrong state %d\n", state_);
+			return false;
+		}
+		return true;
+	}
+
+	bool ReadOutput()
+	{
+		const size_t kChunk = 1024;
+		output_.resize(output_.size() + kChunk);
+		ssize_t n = read(stdout_pipe_, output_.data() + output_.size() - kChunk, kChunk);
+		if (n < 0) {
+			if (errno == EINTR || errno == EAGAIN)
+				return true;
+			fail("proc stdout read failed");
+		}
+		output_.resize(output_.size() - kChunk + n);
+		if (n == 0) {
+			debug("proc %d: output pipe EOF\n", id_);
+			return false;
+		}
+		if (flag_debug) {
+			output_.resize(output_.size() + 1);
+			debug("proc %d: got output: %s\n", id_, output_.data() + debug_output_pos_);
+			output_.resize(output_.size() - 1);
+			debug_output_pos_ = output_.size();
+		}
+		return true;
+	}
+};
+
+// Runner manages a set of test subprocesses (Proc's), receives new test requests from the manager,
+// and dispatches them to subprocesses.
+class Runner
+{
+public:
+	Runner(Connection& conn, const char* name, const char* bin)
+	    : conn_(conn),
+	      name_(name)
+	{
+		size_t num_procs = Handshake();
+		int max_signal_fd = max_signal_ ? max_signal_->FD() : -1;
+		int cover_filter_fd = cover_filter_ ? cover_filter_->FD() : -1;
+		for (size_t i = 0; i < num_procs; i++)
+			procs_.emplace_back(new Proc(conn, bin, i, max_signal_fd, cover_filter_fd,
+						     slowdown_, syscall_timeout_ms_, program_timeout_ms_));
+
+		for (;;)
+			Loop();
+	}
+
+private:
+	Connection& conn_;
+	const char* const name_;
+	std::optional<CoverFilter> max_signal_;
+	std::optional<CoverFilter> cover_filter_;
+	std::vector<std::unique_ptr<Proc>> procs_;
+	std::deque<rpc::ExecRequestRawT> requests_;
+	std::vector<std::string> leak_frames_;
+	uint32 slowdown_ = 0;
+	uint32 syscall_timeout_ms_ = 0;
+	uint32 program_timeout_ms_ = 0;
+
+	friend std::ostream& operator<<(std::ostream& ss, const Runner& runner)
+	{
+		ss << "procs:\n";
+		for (const auto& proc : runner.procs_)
+			ss << *proc;
+		ss << "\nqueued requests (" << runner.requests_.size() << "):\n";
+		for (const auto& req : runner.requests_)
+			ss << req;
+		return ss;
+	}
+
+	void Loop()
+	{
+		Select select;
+		select.Arm(conn_.FD());
+		for (auto& proc : procs_)
+			proc->Arm(select);
+		select.Wait(1000);
```

Comment:
It would help to see a comment re. for what exactly we are waiting here.

Reply:
All these objects contain proper owning containers (std::vector/std::string/std::unique_ptr), so I assume destructors should free everything.

Reply:
I also checked that runner syz-executor memory consumption does not grow as it runs, and fd's are not leaking. After executing 10/20K programs its memory consumption is tiny and steady. After 30K it consumes less than 3MB.

Reply:
I think you've replied to the wrong thread :)

Anyway -- it's good that we don't leak memory, but very bad that it's not entirely clear from the code. I fear that we might unknowingly break it one day.

Reply:
Right.

> It would help to see a comment re. for what exactly we are waiting here.

Added a comment.

> Anyway -- it's good that we don't leak memory, but very bad that it's not entirely clear from the code. I fear that we might unknowingly break it one day.

Memory can well be leaking in Go programs. We set ulimit 255 for open file descriptors, so if we leak any per test, it will explode quickly. For memory, we could write a test (memory shouldn't grow after repeated test executions).